rclpy/rclpy/parameter_client.py

```diff
@@ -0,0 +1,315 @@
+        """
+        client_wait_fns = [
+                self._list_parameter_client.wait_for_service,
+                self._set_parameter_client.wait_for_service,
+                self._get_parameter_client.wait_for_service,
+                self._get_parameter_types_client.wait_for_service,
+                self._describe_parameters_client.wait_for_service,
+                self._set_parameters_atomically_client.wait_for_service,
+        ]
+
+        if timeout_sec is None:
+            return all([fn() for fn in client_wait_fns])
+
+        prev = time.time()
+        for wait_for_service in client_wait_fns:
+            if timeout_sec < 0 or not wait_for_service(timeout_sec):
+                return False
+            timeout_sec -= time.time() - prev
+            prev = time.time()
+        return True
+
+    def list_parameters(
+        self,
+        prefixes: Optional[List[str]] = None,
+        depth: int = 1,
+        callback: Optional[Callable] = None
+    ) -> Future:
+        """
+        List all parameters with given prefixs.
+
+        :param prefixes: List of prefixes to filter by.
+        :param depth: Depth of the parameter tree to list.
+        :param callback: Callback function to call when the request is complete.
+        :return: ``Future`` with the result of the request.
+        """
+        request = ListParameters.Request()
+        if prefixes:
+            request.prefixes = prefixes
+        request.depth = depth
+        future = self._list_parameter_client.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def get_parameters(self, names: List[str], callback: Optional[Callable] = None) -> Future:
+        """
+        Get parameters given names.
+
+        :param names: List of parameter names to get.
+        :param callback: Callback function to call when the request is complete.
+        :return: ``Future`` with the result of the request.
+        """
+        request = GetParameters.Request()
+        request.names = names
+        future = self._get_parameter_client.call_async(request)
+        if callback:
+            future.add_done_callback(callback)
+        return future
+
+    def set_parameters(
+        self,
+        parameters: Sequence[Union[Parameter, ParameterMsg]],
+        callback: Optional[Callable] = None
+    ) -> Future:
+        """
+        Set parameters given a list of parameters.
+
+        The result after the returned future is complete
+        will be of type ``rcl_interfaces.srv.SetParameters.Response``.
+
+        :param parameters: Sequence of parameters to set.
+        :param callback: Callback function to call when the request is complete.
+        :return: ``Future`` with the result of the request.
+        """
+        request = SetParameters.Request()
+        request.parameters = [
+            i.to_parameter_msg() if isinstance(i, Parameter) else i for i in parameters
```

Comment:
tiny nitpick: renaming `i` to `param` is more meaningful.

Reply:
I guess so but it's already an ugly long list comprehension and that would push it _over_ 100 lines and I'd have to break it apart again. In any case it's fairly clear what `i` is

Reply:
Sure, I don't think this is any less readable:

```py
            param.to_parameter_msg()
            if isinstance(param, Parameter) else param
            for param in parameters
```

I think it would also be acceptable to wrap this in a function private to the module (since we use the same logic in a couple places). e.g. `__normalize_to_list_of_parameter_msgs`

I won't block the PR on this comment though.